ICSharpCode.Decompiler/IL/ControlFlow/DetectPinnedRegions.cs

```diff
@@ -614,14 +614,26 @@ bool CreatePinnedRegion(Block block, StLoc stLoc)
 					}
 					Branch br = innerBlock.Instructions.LastOrDefault() as Branch;
 					if (br != null && br.TargetBlock.IncomingEdgeCount == 1
-						&& br.TargetContainer == sourceContainer && reachedEdgesPerBlock[br.TargetBlock.ChildIndex] == 0)
+						&& br.TargetContainer == sourceContainer)
 					{
-						// branch that leaves body.
+						Block unpinBlock = null;
+						if (reachedEdgesPerBlock[br.TargetBlock.ChildIndex] == 0)
+						{
+							unpinBlock = br.TargetBlock;
+						}
+						else if (innerBlock.Instructions.Count == 2 &&
+								innerBlock.Instructions[0].MatchIfInstruction(out _, out var trueInstr) &&
+								trueInstr is Branch trueBr && trueBr.TargetContainer == sourceContainer &&
+								reachedEdgesPerBlock[trueBr.TargetBlock.ChildIndex] == 0)
+						{
+							unpinBlock = trueBr.TargetBlock;
```

Comment:
Yes, I used `== 2` since in my testing there were always two instructions with one being an `if` and the second being the fallthrough `br`. If we were to extend the limit to more than 2 instructions I think it would be necessary to check other instructions for potential outgoing edges which would mean additional logic. though in my testing it seems that the current code is enough.